### PR TITLE
Add method call reference resolver

### PR DESCRIPTION
**Pull request by alekitto**
See #668

### CODE REVIEW

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        $context->markIsResolvingFixture('foo');
+
+        $valueResolverContext = new GenerationContext();
+        $valueResolverContext->markIsResolvingFixture('foo');
+        $valueResolverContext->markAsNeedsCompleteGeneration();
+
+        $dummyProphecy = $this->prophesize(DummyWithGetter::class);
+        $dummyProphecy->getFoo('resolved_argument')
+            ->shouldBeCalled()
+            ->willReturn('resolved_value');
```
Comment (theofidry) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: the semi-column should be on the new line (yeah I know not everyone likes it, but let's be consistent :P)

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                    'resolved_argument',
+                    $newSet = ResolvedFixtureSetFactory::create(new ParameterBag(['ping' => 'pong']))
+                )
+            )
+        ;
+        $valueResolverProphecy
+            ->resolve($reference, $fixture, $newSet, $scope, $valueResolverContext)
+            ->willReturn(
+                new ResolvedValueWithFixtureSet(
+                    $instance = $dummyProphecy->reveal(),
```
Comment (theofidry) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: I prefer the:

```php
/** @var DummyWithGetter $dummy */
$dummy = $dummyProphecy->reveal();
```

convention as you don't have typehint warnings for the services with static analysis tools

File: src/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolver.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        }
+
+        $context->markAsNeedsCompleteGeneration();
+        $fixtureReferenceResult = $this->argumentResolver->resolve($value->getReference(), $fixture, $fixtureSet, $scope, $context);
+        $context->unmarkAsNeedsCompleteGeneration();
+
+        /** @var ResolvedFixtureSet $fixtureSet */
+        list($instance, $fixtureSet) = [$fixtureReferenceResult->getValue(), $fixtureReferenceResult->getSet()];
+
+        if (! method_exists($instance, $functionCall->getName())) {
```
Comment (theofidry) [inline src/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolver.php]: the project convention is to use `false ===` instead

Comment (theofidry) [inline src/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolver.php]: I think this check could be avoided or rather done in the catch block below. That way you only check when there's an error hence there is not performance impact for the regular user, and the for the one who did a mistake it stays as user friendly

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        $resolver = new FixtureMethodCallReferenceResolver($valueResolver);
+        $actual = $resolver->resolve($value, $fixture, $set, $scope, $context);
+
+        $this->assertEquals($expected, $actual);
+
+        $valueResolverProphecy->resolve(Argument::cetera())->shouldHaveBeenCalledTimes(2);
+        $dummyProphecy->getFoo(Argument::cetera())->shouldHaveBeenCalledTimes(1);
+    }
+
+    public function testCatchesThrowablesToThrowResolverException()
```
Comment (theofidry) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: `testThrowsAResolverExceptionOnError`? (error to be taken in the english sense, not the `\Error` object)

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            $resolver->resolve($value, new FakeFixture(), $set, [], new GenerationContext());
+
+            $this->fail('Expected exception to be thrown.');
+        } catch (UnresolvableValueException $exception) {
+            $this->assertEquals(
+                'Could not resolve value "mutable->getFoo()".',
+                $exception->getMessage()
+            );
+            $this->assertEquals(0, $exception->getCode());
+            $this->assertEquals($ex, $exception->getPrevious());
```
Comment (theofidry) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: `$this->assertSame()`

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    {
+        try {
+            $value = new FixtureMethodCallValue(
+                $reference = new MutableValue(new DummyWithGetter()),
+                $functionCall = new FunctionCallValue('getFoo')
+            );
+            $set = ResolvedFixtureSetFactory::create(new ParameterBag(['foo' => 'bar']));
+
+            $dummyProphecy = $this->prophesize(DummyWithGetter::class);
+            $dummyProphecy->getFoo()->willThrow($ex = new \Exception('Inner exception'));
```
Comment (theofidry) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: ~~`$ex = new \Exception('Inner exception')`~~ `$error = new \Error()`

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+                new GenerationContext()
+            );
+
+            $this->fail('Expected exception to be thrown.');
+        } catch (NoSuchMethodException $exception) {
+            $this->assertEquals(
+                'Could not find the method "getNonExistent" of the object "dummy" (class: Dummy).',
+                $exception->getMessage()
+            );
+            $this->assertEquals(0, $exception->getCode());
```
Comment (theofidry) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: there is not previous in that case? (if so I think a `$this->assertNull($exception->getPrevious))` would be appropriate

File: tests/Loader/LoaderIntegrationTest.php [hunk 8]
```diff
@@ -2575,5 +2575,29 @@ public function provideFixturesToGenerate()
                 ],
             ],
         ];
+
+        yield 'method call reference value' => [
```
Comment (theofidry) [inline tests/Loader/LoaderIntegrationTest.php]: could you check if there is a test for a method taking a parameter? If not I think it's worth adding as well in that PR

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        try {
+            $value = new FixtureMethodCallValue(
+                $reference = new MutableValue(new DummyWithGetter()),
+                $functionCall = new FunctionCallValue('getFoo')
+            );
+            $set = ResolvedFixtureSetFactory::create(new ParameterBag(['foo' => 'bar']));
+
+            $error = new \Error();
+            $dummyProphecy = $this->prophesize(DummyWithGetter::class);
+            $dummyProphecy->getFoo()->will(function () use ($error) {
```
Comment (alekitto) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: Can't use `willThrow` here as Prophecy `ThrowPromise` cannot be constructed with Errors

File: tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php (hunk 7 above)
Comment (alekitto) [inline tests/Generator/Resolver/Value/Chainable/FixtureMethodCallReferenceResolverTest.php]: 👍 added

File: tests/Loader/LoaderIntegrationTest.php (hunk 8 above)
Comment (alekitto) [inline tests/Loader/LoaderIntegrationTest.php]: Unit test with argument resolution was present.
I've added a test here with a method taking an argument that is a reference to another object.

**Comment by theofidry** [comment]
Thanks you @alekitto

**Pull request merged**